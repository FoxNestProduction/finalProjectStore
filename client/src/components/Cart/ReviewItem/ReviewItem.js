import React from 'react';
import Card from '@mui/material/Card';
import CardHeader from '@mui/material/CardHeader';
import CardMedia from '@mui/material/CardMedia';
import CardContent from '@mui/material/CardContent';
import CardContentText from '@mui/material/CardContentText';
import CardActions from '@mui/material/CardActions';
import Avatar from '@mui/material/Avatar';

const ReviewItem = () => {
  return (
    <Card>
      <CardHeader>
        <Avatar>
          {avatarUrl}
        </Avatar>
        title={name}
        {/* subheader="Data" */}
      </CardHeader>
      <CardContent>
        <CardContentText>

        </CardContentText>
      </CardContent>

    </Card>
  )
}

export default ReviewItem