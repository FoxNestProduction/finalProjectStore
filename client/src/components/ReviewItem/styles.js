export const stylesCardReview = {
  background:
  'background.paper',
  position: 'relative',
  minWidth: {
    tablet: '295px',
  },
  width: {
    mobile: '100%',
    tablet: '37%',
  },
  px: '9px',
  pt: { mobile: '10px', tablet: '22px' },
  mb: '48px',
  boxShadow: '0px 0px 25px #0000001A',
};

export const stylesQuoteIcon = {
  transform: 'rotate(180deg)',
  color: 'background.quote',
  fontSize: '50px',
};

export const stylesContent = {
  fontSize: {
    mobile: '11px',
    tablet: '13px',
    desktop: '16px',
  },
  fontStyle: 'italic',
  overflow: 'auto',
  height: '45%',
  pb: {
    mobile: 10,
    tablet: 14,
    desktop: 16,
  },
};

export const stylesActionCard = {
  width: '90%',
  p: 2,
  position: 'absolute',
  display: 'flex',
  justifyContent: 'space-between',
  bottom: {
    mobile: '10px',
    tablet: '20px',
    desktop: '30px',
  },
};
