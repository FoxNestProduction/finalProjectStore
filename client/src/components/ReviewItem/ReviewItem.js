import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { useLocation, useNavigate } from 'react-router-dom';
import Card from '@mui/material/Card';
import CardHeader from '@mui/material/CardHeader';
import CardContent from '@mui/material/CardContent';
import CardActions from '@mui/material/CardActions';
import Rating from '@mui/material/Rating';
import Avatar from '@mui/material/Avatar';
import Typography from '@mui/material/Typography';
import FormatQuoteRoundedIcon from '@mui/icons-material/FormatQuoteRounded';
import PersonOutlineOutlinedIcon from '@mui/icons-material/PersonOutlineOutlined';
import MoreHorizIcon from '@mui/icons-material/MoreHoriz';
import { Box } from '@mui/material';
import { searchReviews } from '../../redux/slices/reviewsSlice';

import { stylesCardReview, stylesQuoteIcon, stylesActionCard, stylesContent, stylesText, stylesFullText } from './styles';

const ReviewItem = ({ review }) => {
  const location = useLocation();
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const { rating, content, avatarUrl, userReview, date, _id: id } = review;
  const isMoreThreeLineText = (content.length >= 150) && (location.pathname !== '/reviews');
  const styleComment = (location.pathname === '/reviews') ? stylesFullText : { ...stylesText };
  const widthWrapper = !(location.pathname === '/reviews') ? { mobile: '100%', tablet: '345px', desktop: '485px' } : '100%';

  const ratingNumber = Number(rating);
  const dateReview = new Date(+date);
  const day = String(dateReview.getDate()).padStart(2, '0');
  const month = String(dateReview.getMonth() + 1).padStart(2, '0');
  const year = dateReview.getFullYear();
  const formattedDate = `${day}.${month}.${year}`;

  const handleReviewClick = () => {
    navigate('/reviews');
    dispatch(searchReviews(id));
  };

  return (
    <Card sx={{ ...stylesCardReview, width: widthWrapper }} onClick={() => handleReviewClick(id)}>
      <CardHeader
        avatar={(
          <Avatar>
            { avatarUrl || <PersonOutlineOutlinedIcon /> }
          </Avatar>
        )}
        title={
          userReview
        }
        action={
          <FormatQuoteRoundedIcon sx={stylesQuoteIcon} />
        }
      />
      <CardContent sx={stylesContent}>
        <Box sx={{ ...styleComment }}>
          {content}
        </Box>
        {isMoreThreeLineText && (
          <MoreHorizIcon
            sx={{ position: 'absolute', right: '30px', cursor: 'pointer' }}
          />
        )}
      </CardContent>
      <CardActions sx={stylesActionCard}>
        <Rating
          name="simple-controlled"
          value={ratingNumber}
          readOnly
        />
        <Typography sx={{ color: 'text.secondary' }}>{formattedDate}</Typography>
      </CardActions>
    </Card>
  );
};

ReviewItem.propTypes = {
  review: PropTypes.object,
};
ReviewItem.defaultProps = {
  review: {},
};

export default ReviewItem;
