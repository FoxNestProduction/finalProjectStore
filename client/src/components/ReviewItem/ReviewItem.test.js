import React from 'react';
import { Provider, useDispatch } from 'react-redux';
import { render, screen } from '@testing-library/react';
import { MemoryRouter, useLocation, useNavigate } from 'react-router-dom';

import configureStore from 'redux-mock-store';
import ReviewItem from './ReviewItem';

const mockStore = configureStore();

describe('ReviewItem component', () => {
  const review = {
    _id: '1',
    rating: 4,
    content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    date: '2023-11-09T12:00:00Z',
    customer: {
      lastName: 'Doe',
      firstName: 'John',
    },
  };
  const store = mockStore();

  test('should render ReviewItem', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <ReviewItem review={review} index={0} />
        </MemoryRouter>
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('should render ReviewItem with correct data', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <ReviewItem review={review} index={0} />
        </MemoryRouter>
      </Provider>,
    );

    expect(screen.getByText('Doe John')).toBeInTheDocument();
    expect(screen.getByText('Lorem ipsum dolor sit amet, consectetur adipiscing elit.')).toBeInTheDocument();
    expect(screen.getByText('09.11.2023')).toBeInTheDocument();
  });
});
