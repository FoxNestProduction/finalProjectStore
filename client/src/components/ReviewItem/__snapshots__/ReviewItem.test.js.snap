// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewItem component should render ReviewItem 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-s57lfj-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PersonOutlineOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 5.9c1.16 0 2.1.94 2.1 2.1s-.94 2.1-2.1 2.1S9.9 9.16 9.9 8s.94-2.1 2.1-2.1m0 9c2.97 0 6.1 1.46 6.1 2.1v1.1H5.9V17c0-.64 3.13-2.1 6.1-2.1M12 4C9.79 4 8 5.79 8 8s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 9c-2.67 0-8 1.34-8 4v3h16v-3c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
        >
          Doe John
        </span>
      </div>
      <div
        class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-ddv8by-MuiSvgIcon-root"
          data-testid="FormatQuoteRoundedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.17 17c.51 0 .98-.29 1.2-.74l1.42-2.84c.14-.28.21-.58.21-.89V8c0-.55-.45-1-1-1H5c-.55 0-1 .45-1 1v4c0 .55.45 1 1 1h2l-1.03 2.06c-.45.89.2 1.94 1.2 1.94zm10 0c.51 0 .98-.29 1.2-.74l1.42-2.84c.14-.28.21-.58.21-.89V8c0-.55-.45-1-1-1h-4c-.55 0-1 .45-1 1v4c0 .55.45 1 1 1h2l-1.03 2.06c-.45.89.2 1.94 1.2 1.94z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-szyw0g-MuiCardContent-root"
    >
      <div
        class="MuiBox-root css-jl15ok"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit.
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-h6xg2z-MuiCardActions-root"
    >
      <span
        aria-label="4 Stars"
        class="MuiRating-root MuiRating-sizeMedium Mui-readOnly MuiRating-readOnly css-ryrseu-MuiRating-root"
        role="img"
      >
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarBorderIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        09.11.2023
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ReviewItem component should render ReviewItem, when location "/" 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-s57lfj-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PersonOutlineOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 5.9c1.16 0 2.1.94 2.1 2.1s-.94 2.1-2.1 2.1S9.9 9.16 9.9 8s.94-2.1 2.1-2.1m0 9c2.97 0 6.1 1.46 6.1 2.1v1.1H5.9V17c0-.64 3.13-2.1 6.1-2.1M12 4C9.79 4 8 5.79 8 8s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 9c-2.67 0-8 1.34-8 4v3h16v-3c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
        >
          Doe John
        </span>
      </div>
      <div
        class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-ddv8by-MuiSvgIcon-root"
          data-testid="FormatQuoteRoundedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.17 17c.51 0 .98-.29 1.2-.74l1.42-2.84c.14-.28.21-.58.21-.89V8c0-.55-.45-1-1-1H5c-.55 0-1 .45-1 1v4c0 .55.45 1 1 1h2l-1.03 2.06c-.45.89.2 1.94 1.2 1.94zm10 0c.51 0 .98-.29 1.2-.74l1.42-2.84c.14-.28.21-.58.21-.89V8c0-.55-.45-1-1-1h-4c-.55 0-1 .45-1 1v4c0 .55.45 1 1 1h2l-1.03 2.06c-.45.89.2 1.94 1.2 1.94z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-szyw0g-MuiCardContent-root"
    >
      <div
        class="MuiBox-root css-jl15ok"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. orem ipsum dolor sit amet, consectetur adipiscing elit. orem ipsum dolor sit amet, consectetur adipiscing elit.
      </div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1rs7go6-MuiSvgIcon-root"
        data-testid="MoreHorizIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-h6xg2z-MuiCardActions-root"
    >
      <span
        aria-label="4 Stars"
        class="MuiRating-root MuiRating-sizeMedium Mui-readOnly MuiRating-readOnly css-ryrseu-MuiRating-root"
        role="img"
      >
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarBorderIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
      >
        09.11.2023
      </p>
    </div>
  </div>
</DocumentFragment>
`;
